### PR TITLE
Allow a form with parent selection to be a case list form

### PR DESCRIPTION
**Pull request by snopoke**
@TylerSheffels 

This expands the scope of double management (case list forms) to subcases.

Possibly pull in @dannyroberts 

Probably easiest to review by commit.

### CODE REVIEW

**Comment by dimagimon** [comment]
There are 5 errors:

* corehq/apps/app_manager/models.py, line 2007 - E301 expected 1 blank line, found 0
* corehq/apps/app_manager/suite_xml.py, line 872 - W293 blank line contains whitespace
* corehq/apps/app_manager/suite_xml.py, line 888 - E501 line too long (130 > 115 characters)
* corehq/apps/app_manager/suite_xml.py, line 903 - E303 too many blank lines (2)
* corehq/apps/app_manager/tests/test_suite.py, line 733 - E303 too many blank lines (2)

**Comment by dimagimon** [comment]
There are 5 errors:

* corehq/apps/app_manager/models.py, line 2010 - E301 expected 1 blank line, found 0
* corehq/apps/app_manager/suite_xml.py, line 885 - W293 blank line contains whitespace
* corehq/apps/app_manager/suite_xml.py, line 898 - E301 expected 1 blank line, found 0
* corehq/apps/app_manager/suite_xml.py, line 905 - E501 line too long (126 > 115 characters)
* corehq/apps/app_manager/tests/test_suite.py, line 742 - E303 too many blank lines (2)

File: corehq/apps/app_manager/models.py [hunk 1]
```diff
@@ -1989,18 +1989,55 @@ def requires(self):
...
+            return True
+
+        reg_action = reg_actions[0]
+        parent_tag = reg_action.parent_tag
+        if not parent_tag:
+            return False
+
+        actions_by_tag = deepcopy(self.actions.actions_meta_by_tag)
+        actions_by_tag.pop(reg_action.case_tag)
+        def check_parents(tag):
```
Comment (dimagimon) [inline corehq/apps/app_manager/models.py]: E301 expected 1 blank line, found 0

File: corehq/apps/app_manager/suite_xml.py [hunk 2]
```diff
@@ -870,15 +869,66 @@ def get_if_clause(case_count_xpath):
...
+
+                source_case_id = session_var(source_session_var)
+                case_count = CaseIDXPath(source_case_id).case().count()
+
                 frame_case_created = StackFrameMeta(None, get_if_clause(case_count.gt(0)))
                 frame_case_created.add_child(target_command)
-                frame_case_created.add_child(StackDatum(id=target_session_var, value=source_case_id))
                 stack_frames.append(frame_case_created)
-
+
```
Comment (dimagimon) [inline corehq/apps/app_manager/suite_xml.py]: W293 blank line contains whitespace

File: corehq/apps/app_manager/suite_xml.py [hunk 3]
```diff
@@ -870,15 +869,66 @@ def get_if_clause(case_count_xpath):
...
+                target_form_dm = self.get_form_datums(target_module.get_form(0))
+
+                def get_target_dm(case_type):
+                    try:
+                        [target_dm] = [
+                            target_meta for target_meta in target_form_dm
+                            if target_meta.case_type == case_type
+                        ]
+                    except ValueError:
+                        raise SuiteError("Return module for case list form has mismatching datums: {}".format(form.unique_id))
```
Comment (dimagimon) [inline corehq/apps/app_manager/suite_xml.py]: E501 line too long (126 > 115 characters)

File: corehq/apps/app_manager/tests/test_suite.py [hunk 4]
```diff
@@ -680,6 +689,72 @@ def test_case_list_registration_form_advanced_autoload(self):
...
+        house_module.case_type = 'house'
+        house_module.case_list_form.form_id = register_house_form.unique_id
+        update_house_form = app.new_form(2, 'Update house', lang='en')
+        update_house_form.unique_id = 'update_house_form'
+        update_house_form.requires = 'case'
+        update_house_form.actions.update_case = UpdateCaseAction(update={'question1': '/data/question1'})
+        update_house_form.actions.update_case.condition.type = 'always'
+
+
+        person_module = app.add_module(Module.new_module('Manage person', None))
```
Comment (dimagimon) [inline corehq/apps/app_manager/tests/test_suite.py]: E303 too many blank lines (2)

**Comment by snopoke** [comment]
I'm going to test this on a phone tomorrow.

**Comment by dannyroberts** [comment]
@snopoke is this ready for me to be reviewing?

**Comment by snopoke** [comment]
Yes, please review. Let me know if you want to chat though it. There is
some refactoring I want to do in the WorkflowHelper class but don't want to
muddy the water any more for this PR.

**Comment by dannyroberts** [comment]
I did my pass on review, PR looks pretty heroic. Noticed during the review that it takes quite a bit of effort for me to understand the changes on any given line, which since I spent years of my life working on this code is a sign that things are getting pretty complicated in there. Wondering whether you feel the same way or I just need to get used to it again, and whether it's affected your ability to work easily in the code / is something we should be flagging/thinking about.

**Comment by snopoke** [comment]
@dannyroberts thanks for the review! Yes, this code is getting pretty complex - feels more like a compiler every time I work on it. I have some refactors planned e.g. https://github.com/dimagi/commcare-hq/pull/7761 but would really appreciate your thoughts on ways to simplify things and make it easier to work in.

**Comment by dannyroberts** [comment]
Ok all your answers look good to me, so consider my review complete. You've got the hold off on merge label set, so I won't merge, but feel free to merge if you're ready.

**Comment by TylerSheffels** [comment]
I took a quick pass, just one question.  Looks good to my untrained eye.

**Comment by snopoke** [comment]
Ran the app builder live tests on this and the output is the same as master. Did find one bug though fixed here: https://github.com/dimagi/commcare-hq/commit/73e60c192d21d7e131518f6ed7e29dc7a74060bc

Will merge once tests pass.

**Comment by dannyroberts** [comment]
:+1: on the app builder live tests. I always feel better after running that :)

**Pull request merged**